packages/circuits-sdk/src/implementations/CircomWrapper.ts

```diff
@@ -1,29 +1,108 @@
-import { ObjectUtils } from "@snickerdoodlelabs/common-utils";
-import { CircuitError, JSONString, ZKProof } from "@snickerdoodlelabs/objects";
+import { ObjectUtils, IAxiosAjaxUtils } from "@snickerdoodlelabs/common-utils";
+import {
+  CircuitError,
+  JSONString,
+  ZKProof,
+  IpfsCID,
+  AjaxError,
+} from "@snickerdoodlelabs/objects";
 import { injectable, unmanaged } from "inversify";
-import { ResultAsync } from "neverthrow";
+import { ResultAsync, errAsync, okAsync } from "neverthrow";
 import { CircuitSignals, Groth16Proof, PublicSignals, groth16 } from "snarkjs";
+import { urlJoin } from "url-join-ts";
 
+import { ICircuitsSDKConfigProvider } from "@circuits-sdk/ICircuitsSDKConfigProvider.js";
+import { ICircomWrapper } from "@circuits-sdk/interfaces/ICircomWrapper.js";
 @injectable()
 export abstract class CircomWrapper<
   TInput extends CircuitSignals,
   TVerifyInput extends PublicSignals,
-> {
+> implements ICircomWrapper
+{
+  private _wasmResult?: ResultAsync<Uint8Array, AjaxError | CircuitError>;
+  private _zkeyResult?: ResultAsync<Uint8Array, AjaxError | CircuitError>;
   public constructor(
-    @unmanaged() protected wasm: Uint8Array,
-    @unmanaged() protected zkey: Uint8Array,
+    @unmanaged() protected ajaxUtils: IAxiosAjaxUtils,
+    @unmanaged() protected circuitsSDKConfig: ICircuitsSDKConfigProvider,
     @unmanaged() protected vkey: Record<string, unknown>,
+    @unmanaged() protected wasmKey: IpfsCID,
+    @unmanaged() protected zkeyKey: IpfsCID,
   ) {}
 
+  public preFetch(): ResultAsync<undefined, never> {
+    return this.circuitsSDKConfig.getConfig().andThen((config) => {
+      const wasmUrl = new URL(
+        urlJoin(config.circuitsIpfsFetchBaseUrl, this.wasmKey),
+      );
+      const zkeyUrl = new URL(
+        urlJoin(config.circuitsIpfsFetchBaseUrl, this.zkeyKey),
+      );
+
+      this._wasmResult = this.fetchResource(wasmUrl);
+      this._zkeyResult = this.fetchResource(zkeyUrl);
+
+      return okAsync(undefined);
+    });
+  }
+
+  private fetchResource(
+    url: URL,
+  ): ResultAsync<Uint8Array, AjaxError | CircuitError> {
+    return this.ajaxUtils
+      .get<string>(url)
+      .map((data) => {
+        return new Uint8Array(Buffer.from(data, "base64").buffer);
+      })
+      .mapErr(
+        (error) =>
+          new CircuitError(`Error fetching resource from ${url}`, error),
+      );
+  }
+
+  private getResourceUrl(key: IpfsCID): ResultAsync<URL, CircuitError> {
```

Comment:
Why do you need fetchResource() and getResourceURL? You aren't using this method in preFetch() above.

Reply:
I would make fetchResource() just take the IPFS CID.

Reply:
that one is for wasmResult and zkeyResult getters